### PR TITLE
fix: Enhance SelectWidget label and value handling logic

### PR DESCRIPTION
**Pull request by rahulbarwal**
## Description

- Updated the label extraction logic to support both static and dynamic keys from sourceData.
- Improved value mapping to handle cases where optionValue is an array, ensuring consistency in label and value retrieval.
- Added detailed documentation comments to clarify the expected structure of sourceData and the behavior of optionLabel and optionValue props.

Fixes #`Issue Number`  
_or_  
Fixes `Issue URL`
> [!WARNING]  
> _If no issue exists, please create an issue first, and check with the maintainers if the issue is valid._

## Automation

/ok-to-test tags="@tag.Select"

### :mag: Cypress test results

> [!TIP]
> 🟢 🟢 🟢 All cypress tests have passed! 🎉 🎉 🎉
> Workflow run: <https://github.com/appsmithorg/appsmith/actions/runs/12409964844>
> Commit: 1027f13627983a05594eb85d63ea8ecd06c6b40e
> <a href="https://internal.appsmith.com/app/cypress-dashboard/rundetails-65890b3c81d7400d08fa9ee5?branch=master&workflowId=12409964844&attempt=1" target="_blank">Cypress dashboard</a>.
> Tags: `@tag.Select`
> Spec:
> <hr>Thu, 19 Dec 2024 09:59:56 UTC

## Communication
Should the DevRel and Marketing teams inform users about this change?
- [ ] Yes
- [ ] No

## Summary by CodeRabbit

- **New Features**
	- Enhanced logic for handling `optionLabel` and `optionValue` properties in the `SelectWidget` component, allowing for more flexible input scenarios.
	- Improved validation functions to handle cases where input values are arrays with identical elements, enabling direct parsing from source data.

### CODE REVIEW

File: app/client/src/widgets/SelectWidget/widget/derived.js [hunk 1]
```diff
@@ -5,16 +5,57 @@
...
+      const allValuesEqual = props.optionValue.every(
+        (value, _, arr) => value === arr[0],
+      );
+      const doesKeyExistInSourceData = props.optionValue[0] in sourceData[0];
+
+      if (allValuesEqual && doesKeyExistInSourceData) {
+        values = sourceData.map((d, i) => d[props.optionValue[0]]);
+      } else {
+        values = props.optionValue;
+      }
```
[the remark this answers is not in the document]
Comment (rahulbarwal) [reply app/client/src/widgets/SelectWidget/widget/derived.js]: thanks for great tip.

File: app/client/src/widgets/SelectWidget/widget/derived.js [hunk 2]
```diff
@@ -5,16 +5,57 @@ export default {
...
+      const allLabelsEqual = props.optionLabel.every(
+        (label, _, arr) => label === arr[0],
+      );
+      const doesKeyExistInSourceData = props.optionLabel[0] in sourceData[0];
+
+      if (allLabelsEqual && doesKeyExistInSourceData) {
+        labels = sourceData.map((d, i) => d[props.optionLabel[0]]);
+      } else {
+        labels = props.optionLabel;
+      }
```
[the remark this answers is not in the document]
Comment (rahulbarwal) [reply app/client/src/widgets/SelectWidget/widget/derived.js]: done.

**Review by jsartisan — APPROVED** [review]

**Pull request merged**